### ISSUE TITLE
Bug: [no-invalid-void-type] conditional return type false positive

### ISSUE THREAD

**Issue by panmenghan** [typescript-eslint/typescript-eslint#5192]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=GYVwdgxgLglg9mABCAzgUwMIEN0B4AqAfABRQBci+ANIhFgDb0BGWEA1hcQJSIC8hlRGgAeUNGAAmKZJLTAYYNBMQB+RADc4MZRRRQATgoDmPAN4BYAFCJaCFHHpoAdPThHSNOoxbsuVgL5WVqiYOGjEAIw03HwCAOQRcX6WVgD0qYgADjjSAAbgEnIKSrmIUHCIDK4A7ohgcFCI+mgAjiAwzWUVzVAg+khYGgwgaMHo2OjEBUWKEtE8-IimiP48QZZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWMBuAhvQCbUubcoILF0URNGhtokAL4gFQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function useCase<T>(t: T, callback: () => T extends undefined ? void : string) {
  console.log(t, callback)
}

useCase(1, () => '1')

// pass `undefined` to allow not require to return a value
useCase(undefined, () => { })
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "rules": {"@typescript-eslint/no-invalid-void-type": "error"}
  },
};
```

### Expected Result

no error

### Actual Result

void is only valid as a return type or generic type variable. 1:65 - 1:69

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.28.0` |
| `@typescript-eslint/parser`            | `5.28.0` |
| `TypeScript`                           | `4.7.3` |
| `ESLint`                               | `8.17.0` |
| `node`                                 | `16.15.1` |

**Comment by bradzacher** [comment]
I'll be honest when I say that your usecase is very strange. I don't think I've ever seen someone use a conditional type to conditionally type a function return type as `void`.

I think we can add support for this.

**Comment by panmenghan** [comment]
I try to do something like this:

```ts
// client-side-lib.ts
type Input = 0
type Output = 1
declare global {
  interface RpcSchema { // <- will be augmented by app
    // [key: string]: [input: unknown, output: unknown]
  } 
}
export async function invokeRpc<RpcName extends keyof RpcSchema>(
  rpcName: RpcName,
  input: RpcSchema[RpcName][Input],
) {
  const output = await sendRpcRequest(rpcName, input)
  return output as RpcSchema[RpcName][Output]
}

// server-side-lib.ts
type Input = 0
type Output = 1
declare global {
  interface RpcSchema { // <- will be augmented by app
    // [key: string]: [input: unknown, output: unknown] 
  } 
}
export function exposeRpc<RpcName extends keyof RpcSchema>(
  rpcName: RpcName,
  callback: (input: RpcSchema[RpcName][Input]) => Promise<
    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
    RpcSchema[RpcName][Output] extends undefined ? void : RpcSchema[RpcName][Output] // <- the case
  >
) {
  saveRpcCallback(rpcName, callback)
}

// shared-interface-app.ts
declare global {
  // augment both client and server side schemas with the following fields:
  interface RpcSchema {
    'someTable.newOne': [input: {col1: string, ...}, output: {id: string, ...}]
    'someTable.deleteOne': [input: {id: string}, output: undefined] // <- the case related    
    // maybe should change `undefined` to `void` here,
    // so instead of doing conditional return type in `exposeRpc`,
    // just disable the `@typescript-eslint/no-invalid-void-type` rule globally,
    // and let `tsc` to check the invalid void type error?
  }
}

// client-side-app.ts
import {invokeRpc} from 'client-side-lib'
const newRow = await invokeRpc('someTable.newOne', {col1: 'foo', ...})

invokeRpc('someUnknownRpc', {}) // <- expected error: Argument of type '"someUnknownRpc"' is not assignable to parameter of type 'keyof RpcSchema'.

// server-side-app.ts
import {exposeRpc} from 'server-side-lib'
exposeRpc('someTable.newOne', async input => {
  return {id: '1', ...}
})

exposeRpc('someUnknownRpc', () => {}) // <- expected error: Argument of type '"someUnknownRpc"' is not assignable to parameter of type 'keyof RpcSchema'.
```

**Comment by Zamiell** [comment]
See this (tiny) playground : https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=PTAEDkHtQdwQwE4DsCWSDmoAWBTBOA6AWACgATHAYwBtEdQAzAVyUoBcVIlHJIAKAJQAuUACNe1HHCQBuUqRCgA6olQZseHABpYWOG1AoAzqHSQ0mLgH4bpCjTqMW7Tt1GJBI8ZEnTQAH1AAN3MyORIgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KAHc+0JpHBgAviDVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

Is this this playground bug the same as the this issue, or is it different? If the latter, I'll open a new issue.

**Comment by aqueenan** [comment]
If anyone needs a workaround in the meantime, try this:

```typescript
type Void<V = void> = V
function useCase<T>(t: T, callback: () => T extends undefined ? Void : string) {
  console.log(t, callback)
}
```

**Comment by Josh-Cena** [comment]
I'm going to close this as a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/8113. Let's discuss there instead!
